## Final Project Submission

Please fill out:
* Student name: Dahir Ahmed
* Student pace: Part time 
* Scheduled project review date/time:27TH OCT TO  31ST OCT 2025
* Instructor name: 
* Blog post URL:


In [10]:
# Your code here - remember to use markdown cells for comments as well!
# Import the pandas library for data manipulation and analysis
import pandas as pd

# Import the numpy library for numerical operations and working with arrays
import numpy as np


# Import the warnings module to control or suppress warning messages
import warnings

# Suppress all warning messages to keep the output clean
warnings.filterwarnings('ignore')



1.1) Read Aviation_Data.csv into a pandas DataFrame named df

In [14]:
# Load the CSV file "Aviation_Data.csv" from the 'data' folder into a pandas DataFrame named 'df'

df = pd.read_csv("./data/Aviation_Data.csv")

# Display the first 5 rows of the DataFrame to quickly inspect the data

df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country Latitude Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States      NaN       NaN          NaN   
1   BRIDGEPORT, CA  United States      NaN       NaN          NaN   
2    Saltville, VA  United States  36.9222  -81.8781          NaN   
3       EUREKA, CA  United States      NaN       NaN          NaN   
4       Canton, OH  United States      NaN       NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In [19]:
# Display the shape (rows, columns)
df.shape

# Print the number of rows and columns in a readable format
print(f"This dataset has {df.shape[0]} rows and {df.shape[1]} columns.")


This dataset has 90348 rows and 31 columns.


In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

# How to Find Duplicates in our Aviation Data 

our data has **90,348 rows** and **31 columns**.  
Sometimes the **same crash** gets entered **twice** by mistake.  
We want to **find & remove** those duplicates so our analysis is clean!

we have 88889 event ids(which should be unique) but invistagation type 90348 techically we have duplicates of 
(90348-888889) 1489

In [43]:
df['Event.Id'].value_counts()

20001212X19172    3
20001214X45071    3
20001213X33965    2
20001208X07187    2
20020214X00224    2
                 ..
20001213X34057    1
20020917X04401    1
20110418X31323    1
20001213X29614    1
20221011106090    1
Name: Event.Id, Length: 87951, dtype: int64

In [30]:
#Check missing data percentages:
df.isnull().mean().sort_values(ascending=False) * 100 


Schedule                  86.073848
Air.carrier               81.573471
FAR.Description           64.555939
Aircraft.Category         64.263736
Longitude                 61.954886
Latitude                  61.944924
Airport.Code              44.382831
Airport.Name              41.570372
Broad.phase.of.flight     31.681941
Publication.Date          18.471909
Total.Serious.Injuries    15.461327
Total.Minor.Injuries      14.822686
Total.Fatal.Injuries      14.233851
Engine.Type                9.447913
Report.Status              8.677558
Purpose.of.flight          8.468367
Number.of.Engines          8.348829
Total.Uninjured            8.158454
Weather.Condition          6.586753
Aircraft.damage            5.150086
Registration.Number        3.072564
Injury.Severity            2.721698
Country                    1.865011
Amateur.Built              1.727764
Model                      1.716695
Make                       1.684597
Location                   1.672422
Event.Date                 1

In [29]:
df.describe()


Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

In [ ]:
#Get summary stats

df.describe(include='all')

Event.Id Investigation.Type Accident.Number  Event.Date  \
count            88889              90348           88889       88889   
unique           87951                 71           88863       14782   
top     20001212X19172           Accident      WPR22FA309  1982-05-16   
freq                 3              85015               2          25   
mean               NaN                NaN             NaN         NaN   
std                NaN                NaN             NaN         NaN   
min                NaN                NaN             NaN         NaN   
25%                NaN                NaN             NaN         NaN   
50%                NaN                NaN             NaN         NaN   
75%                NaN                NaN             NaN         NaN   
max                NaN                NaN             NaN         NaN   

             Location        Country Latitude Longitude Airport.Code  \
count           88837          88663    34382     34373        50249   
unique          27758            219    25592     27156        10375   
top     ANCHORAGE, AK  United States  332739N  0112457W         NONE   
freq              434          82248       19        24         1488   
mean              NaN            NaN      NaN       NaN          NaN   
std               NaN            NaN      NaN       NaN          NaN   
min               NaN            NaN      NaN       NaN          NaN   
25%               NaN            NaN      NaN       NaN          NaN   
50%               NaN            NaN      NaN       NaN          NaN   
75%               NaN            NaN      NaN       NaN          NaN   
max               NaN            NaN      NaN       NaN          NaN   

       Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
count         52790  ...             82697       16648         77488.000000   
unique        24871  ...                26       13590                  NaN   
top         Private  ...          Personal       Pilot                  NaN   
freq            240  ...             49448         258                  NaN   
mean            NaN  ...               NaN         NaN             0.647855   
std             NaN  ...               NaN         NaN             5.485960   
min             NaN  ...               NaN         NaN             0.000000   
25%             NaN  ...               NaN         NaN             0.000000   
50%             NaN  ...               NaN         NaN             0.000000   
75%             NaN  ...               NaN         NaN             0.000000   
max             NaN  ...               NaN         NaN           349.000000   

       Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
count            76379.000000         76956.000000    82977.000000   
unique                    NaN                  NaN             NaN   
top                       NaN                  NaN             NaN   
freq                      NaN                  NaN             NaN   
mean                 0.279881             0.357061        5.325440   
std                  1.544084             2.235625       27.913634   
min                  0.000000             0.000000        0.000000   
25%                  0.000000             0.000000        0.000000   
50%                  0.000000             0.000000        1.000000   
75%                  0.000000             0.000000        2.000000   
max                161.000000           380.000000      699.000000   

       Weather.Condition  Broad.phase.of.flight   Report.Status  \
count              84397                  61724           82508   
unique                 4                     12           17007   
top                  VMC                Landing  Probable Cause   
freq               77303                  15428           61754   
mean                 NaN                    NaN             NaN   
std                  NaN                    NaN             NaN   
min          

### 1. Columns to **DROP** (too many blanks + not very helpful)

| Column                  | % Missing | Why drop it? |
|-------------------------|-----------|--------------|
| **Schedule**            | 86%       | Almost all rows are blank. It just says if the flight was on a regular schedule (like a bus). We already know from `Purpose.of.flight` if it was a normal passenger trip. |
| **Air.carrier**         | 82%       | Mostly empty. It's just the airline name (like "Delta"). We don’t need the name to understand the crash. |
| **FAR.Description**     | 65%       | A boring legal rule number. Missing for most small planes. Doesn’t help us predict injuries or damage. |
| **Aircraft.Category**   | 64%       | Says if it's a plane, helicopter, etc. But we already have `Make` and `Model` — that’s enough! |
| **Longitude / Latitude**| ~62%      | GPS numbers. Great for Google Maps, but more than half are missing. We still have city and airport to know *where*. |
| **Airport.Code**        | 44%       | The short code like "LAX" or "JFK". Many crashes happen **away** from airports, so it’s often blank. |
| **Airport.Name**        | 42%       | Same as above. We can just use the city name instead. |
| **Broad.phase.of.flight**| 32%    | Tells you *when* in the flight it crashed (takeoff? landing?). Helpful, but too many blanks and too many options. Hard to guess the missing ones. |
| **Publication.Date**    | 18%       | This is the day the **report** came out — not the crash day. Totally useless for understanding the accident. |

---



In [31]:
cols_to_drop = [
    'Schedule', 'Air.carrier', 'FAR.Description', 'Aircraft.Category',
    'Longitude', 'Latitude', 'Airport.Code', 'Airport.Name',
    'Broad.phase.of.flight', 'Publication.Date'
]

df = df.drop(columns=cols_to_drop)

In [32]:
df.shape

(90348, 21)

In [34]:
df.isnull().mean().sort_values(ascending=False) * 100 

Total.Serious.Injuries    15.461327
Total.Minor.Injuries      14.822686
Total.Fatal.Injuries      14.233851
Engine.Type                9.447913
Report.Status              8.677558
Purpose.of.flight          8.468367
Number.of.Engines          8.348829
Total.Uninjured            8.158454
Weather.Condition          6.586753
Aircraft.damage            5.150086
Registration.Number        3.072564
Injury.Severity            2.721698
Country                    1.865011
Amateur.Built              1.727764
Model                      1.716695
Make                       1.684597
Location                   1.672422
Event.Date                 1.614867
Accident.Number            1.614867
Event.Id                   1.614867
Investigation.Type         0.000000
dtype: float64

### 2. Columns to **KEEP** (a little blank is okay — they’re super important!)

| Column | % Missing | Why keep it? |
|--------|-----------|--------------|
| **Total.Serious.Injuries** | 15% | This tells us how many people got **really hurt**. Most crashes have 0 — so we can just fill blanks with **0**. |
| **Total.Minor.Injuries** | 15% | Same idea — small cuts or bruises. Fill blanks with **0**. |
| **Total.Fatal.Injuries** | 14% | **Super important!** How many people died. We *must* keep this. Fill blanks with **0**. |
| **Engine.Type** | 9% | Is it a normal engine or a jet? Big difference in crashes. Easy to fill with “Unknown” if missing. |
| **Report.Status** | 9% | Says if the report is “final” or “still being checked”. Helps us know if the data is ready. |
| **Purpose.of.flight** | 8% | Was it a normal passenger flight? Training? Crop dusting? **Tells us a lot** about the risk. |
| **Number.of.Engines** | 8% | 1 engine or 2+? Two engines = safer usually. Easy to guess if missing. |
| **Total.Uninjured** | 8% | How many people walked away fine. Completes the story! Fill with a number if missing. |
| **Weather.Condition** | 7% | Was it sunny or stormy? Weather causes many crashes. Most are “good weather” — so fill with that. |
| **Aircraft.damage** | 5% | Did the plane get scratched or totally destroyed? **Main thing** we want to predict. |
| **Registration.Number** | 3% | The plane’s license plate (like N123AB). Helps find duplicates. |
| **Injury.Severity** | 3% | Says “Fatal” or “Minor” in one word. We can make it from the injury counts if needed. |
| **Country, Amateur.Built, Model, Make, Location, Event.Date, Accident.Number, Event.Id** | ≤ 2% | Almost no blanks! These are **must-haves**: 

